src/openrct2/scripting/bindings/ride/ScTrackSegment.cpp

```diff
@@ -141,4 +144,25 @@ DukValue ScTrackSegment::elements_get() const
     return DukValue::take_from_stack(ctx);
 }
 
+uint16_t ScTrackSegment::getSubpositionLength(uint8_t trackSubposition, uint8_t direction) const
+{
+    return vehicle_get_move_info_size(static_cast<VehicleTrackSubposition>(trackSubposition), _type, direction);
```

Comment:
why would you need this you are getting a vector of the whole thing so you don't need this.

Reply:
Because I think getting the whole vector may be less performant if you don't need the whole vector. (Since it's also creating a copy of all that data) The implementations for [tile elements](https://github.com/OpenRCT2/OpenRCT2/blob/7f29e4e39c76dbf7039162368ec8bb604e943d03/distribution/openrct2.d.ts#L792-L795) and [rides & entities](https://github.com/OpenRCT2/OpenRCT2/blob/7f29e4e39c76dbf7039162368ec8bb604e943d03/distribution/openrct2.d.ts#L616-L623) have the same functionality.

Reply:
Would anything ever need that though? rides / entities is makes sense as you often want to quickly know how many rides exist. Why would you care about the length of the subposition array and why would you query it often?

Reply:
It's the use case for why I made the PR to begin with:
> I've been trying to use it to calculate trackProgress-distance between two vehicles. For this I need to know how long track pieces can be, and while there is [`TrackSegment.length`](https://github.com/OpenRCT2/OpenRCT2/blob/121afbe1d0eb1ff526b223468fc887bf8919e954/distribution/openrct2.d.ts#L1241), this is just an arbitrary number for each track piece and not an exact upper bound for `trackProgress`.

Reply:
There is no API for getting a single sub position though, so how are you avoiding getting the whole vector?

Reply:
@IntelOrca for just distance between two vehicles progress-wise you don't really need the actual positions. If car 1 is at trackprogress 5 out of 32 for a flat track piece, and car 2 is at trackprogress 12 out of 32 for the flat track piece after it, then their trackprogress distance is (32 - 12) + 5 = 25 between car 2 and car 1. For this calculation I don't need the subposition, but I do need the length of the vector.

In short, it's distance in progress, not xyz based distance.

Reply:
But feel free to suggest ideas for better API endpoints!

Reply:
Oh, so length does not mean the number of elements in an array, it is an actual metric within the game?

Reply:
@IntelOrca The current `TrackSegment.length` draws its numbers from [this table](https://github.com/OpenRCT2/OpenRCT2/blob/a68f8e84678f1f92c7a6c1bf9410c96bb1e0d00d/src/openrct2/ride/TrackData.cpp#L2820) which is only used for calculating the ride length statistic. While some lengths there are the same as the subposition lengths (flat straight pieces), others are not (like turns).

Reply:
For `getSubpositionLength` it does mean the array length. I added it to prevent pulling in the whole subposition vector if only the length is needed, similar to the API's for [tile elements](https://github.com/OpenRCT2/OpenRCT2/blob/7f29e4e39c76dbf7039162368ec8bb604e943d03/distribution/openrct2.d.ts#L792-L795) and [rides & entities](https://github.com/OpenRCT2/OpenRCT2/blob/7f29e4e39c76dbf7039162368ec8bb604e943d03/distribution/openrct2.d.ts#L616-L623).

But if it is preferred to have it removed, let me know.

Reply:
Can you send me a snippet of a script where you use just the array length?

Reply:
@IntelOrca Here's a Typescript snippet that gets the distance between a car and its preceding car (up to 5 track pieces away, if it has any).

```ts
const MaxForwardIterations = 5;


export function getSpacingToPrecedingVehicle(car: Car): number | null
{
	const precedingCarId = car.previousCarOnRide;
	if (precedingCarId === null)
	{
		return null;
	}

	const carInFront = <Car>map.getEntity(precedingCarId);

	const currentProgress = car.trackProgress;
	const inFrontProgress = carInFront.trackProgress;
	const currentTrackLocation = car.trackLocation;
	const inFrontTrackLocation = carInFront.trackLocation;

	if (isLocationEqual(currentTrackLocation, inFrontTrackLocation))
	{
		// Both cars are on the same track piece
		return (inFrontProgress - currentProgress);
	}

	const iterator = getTrackIteratorAtLocation(currentTrackLocation);
	if (!iterator)
	{
		return null;
	}

	let iteratorSegment = iterator.segment;
	if (!iteratorSegment)
	{
		console.log(`Track iterator does not have segment for car at position; ${currentTrackLocation.x}, ${currentTrackLocation.y}, ${currentTrackLocation.z}, direction; ${currentTrackLocation.direction}`);
		return null;
	}

	// Iterate through the track pieces to find the preceding vehicle.
	const subposition = car.subposition;
	const trackLength = iteratorSegment.getSubpositionLength(subposition, currentTrackLocation.direction);
	let totalDistance = (trackLength - currentProgress);
	for (let i = 0; i < MaxForwardIterations; i++)
	{
		iterator.next();

		const iteratorPosition = iterator.position;
		if (isLocationEqual(iteratorPosition, inFrontTrackLocation))
		{
			// Preceding vehicle found, return total distance between the two.
			return (totalDistance + inFrontProgress);
		}

		iteratorSegment = iterator.segment;
		if (iteratorSegment === null)
		{
			// End of track, look no further.
			return null;
		}

		// Add distance of current track piece and check the next one.
		totalDistance += iteratorSegment.getSubpositionLength(subposition, iteratorPosition.direction);
	}
	return null;
}


function getTrackIteratorAtLocation(trackLocation: CoordsXYZD): TrackIterator | null
{
	const currentTrackIndex = getIndexForTrackElementAt(trackLocation);
	if (currentTrackIndex === null)
	{
		console.log(`Could not find track for car at position; ${trackLocation.x}, ${trackLocation.y}, ${trackLocation.z}, direction; ${trackLocation.direction}`);
		return null;
	}

	const iterator = map.getTrackIterator(trackLocation, currentTrackIndex);
	if (!iterator)
	{
		console.log(`Could not start track iterator for car at position; ${trackLocation.x}, ${trackLocation.y}, ${trackLocation.z}, direction; ${trackLocation.direction}, index; ${currentTrackIndex}`);
		return null;
	}
	return iterator;
}


function getIndexForTrackElementAt(coords: CoordsXYZD): number | null
{
	const tile = map.getTile(coords.x / 32, coords.y / 32);
	const allElements = tile.elements, len = allElements.length;

	for (let i = 0; i < len; i++)
	{
		const element = tile.elements[i];
		if (element.type === "track"
			&& element.baseZ === coords.z
			&& element.direction === coords.direction)
		{
			return i;
		}
	}
	return null;
}


function isLocationEqual(left: CoordsXYZD, right: CoordsXYZD): boolean
{
	return left.x === right.x
		&& left.y === right.y
		&& left.z === right.z
		&& left.direction === right.direction;
}
```

Reply:
Cool, that looks fine